### ISSUE TITLE
missing filetype grass_fermented

### ISSUE THREAD

**Issue by HiPhi77** [Mmtrx/FS22_BetterContracts#105]
Today I got in my log a missing filetype grass_fermented error

some more info: BGA Gutshof is a Pumps N Hoses BGA Fermenter
I am playing with maize+ enabled (but to my knowledge, maize+ does not touch the PnH BGAs, just adds a fill type converter)
map (should be irrelevant) Ostseeküste
no other errors in the log
I assume better contracts want to generate a new missing with my own BGA as delivery point.

2023-02-17 17:17  Missing filltype GRASS_FERMENTED BGA Gutshof: Fermenter 20m
2023-02-17 17:17 LUA call stack:
  =dataS/scripts/objects/SellingStation.lua (430) : printCallstack
  =/Users/philipp/Library/Application Support/FarmingSimulator2022/mods/FS22_BetterContracts/betterContracts.lua (746) : getFilltypePrice
  =/Users/philipp/Library/Application Support/FarmingSimulator2022/mods/FS22_BetterContracts/betterContracts.lua (649) : addMission
  =/Users/philipp/Library/Application Support/FarmingSimulator2022/mods/FS22_BetterContracts/scripts/gui.lua (310) : refresh
  =dataS/scripts/gui/InGameMenuContractsFrame.lua (154) : updateList
  =/Users/philipp/Library/Application Support/FarmingSimulator2022/mods/FS22_BetterContracts/scripts/gui.lua (168) : superFunc
  =dataS/scripts/gui/base/TabbedMenu.lua (614) : onFrameOpen
  =dataS/scripts/internalMods/FS22_precisionFarming/scripts/gui/InGameMenuExtension.lua (144) : superFunc
  =dataS/scripts/gui/InGameMenu.lua (852) : onPageChange
  =dataS/scripts/gui/elements/PagingElement.lua (90) : raiseCallback
  =dataS/scripts/gui/base/TabbedMenu.lua (551) : setPage
  =dataS/scripts/gui/elements/MultiTextOptionElement.lua (424) : raiseCallback
  =dataS/scripts/gui/elements/MultiTextOptionElement.lua (142) : raiseClickCallback
  =dataS/scripts/gui/base/TabbedMenu.lua (252) : setState
  =dataS/scripts/gui/elements/ButtonElement.lua (762) : raiseCallback
  =dataS/scripts/gui/elements/ButtonElement.lua (475) : sendAction
  =dataS/scripts/gui/elements/GuiElement.lua (612) : mouseEvent
  =dataS/scripts/gui/elements/ListItemElement.lua (141) : mouseEvent
  =dataS/scripts/gui/elements/GuiElement.lua (612) : mouseEvent
  =dataS/scripts/gui/elements/ListElement.lua (657) : mouseEvent
  =dataS/scripts/gui/elements/GuiElement.lua (612) : mouseEvent
  =dataS/scripts/gui/elements/GuiElement.lua (612) : mouseEvent
  =dataS/scripts/gui/base/Gui.lua (636) : mouseEvent
  =dataS/scripts/events.lua (26) : mouseEvent
2023-02-17 17:17 Error: Running LUA method 'mouseEvent'.
dataS/scripts/objects/SellingStation.lua:445: attempt to perform arithmetic on a nil value

**Comment by HiPhi77** [comment]
after using the search function, I found 
Mod conflict #95
to describe the same error.
Also did some. testing by using something like
local status,price = pcall(m.sellPoint.getEffectiveFillTypePrice,m.sellPoint,m.fillType)
		if not status then 
			price = 0
		end
this seem to work in gui.lua (but still writes quite a lot of errors in the log)
but when using in better contracts.lue the error repeats again and again.

at least: by using call in gui.lua and hardcoding price 0 0 in bettercontracst.lua (for bales) I can use better contracts.
Not a real solution.

maybe will do some more tests.
I think the issue is the combination of maize+ and PnH BGAs. So if the script could detect a maize+ fill_type and a PnH BGA, it might be able to avoid this only for this combination.

**Comment by Mmtrx** [comment]
Yep, that's an incompatibility with M+. Following fix will be in my next update, you can try it right away:
```
	-- check for Maize+ filltype
	local fillType = m.fillType
	if fillType == FillType.GRASS_FERMENTED then
		filltype = FillType.SILAGE
	end
	return m.sellPoint:getEffectiveFillTypePrice(fillType)
```
changed in betterContracts.lua, function `BetterContracts:getFilltypePrice(m)`

**Comment by HiPhi77** [comment]
tested this and (after aligning fillType camel casing) it works.
but has to be added to gui.lua  too (in line 599 it also checks the price)
local text, text4a, text4b
		--get current price
		--check for Maize+ filltype
		local fillType = m.fillType
		if fillType == FillType.GRASS_FERMENTED then
			fillType = FillType.SILAGE
		end
		local price = m.sellPoint:getEffectiveFillTypePrice(fillType)
		self.my.filltype:setText(c.ftype)

thanks a lot for the support!

**Comment by Mmtrx** [comment]
Ah - I didn't see the filltype typo, because only tested with standard ingame BGA. Those are not a problem, since they contain a price for FillType.GRASS_FERMENTED. Must be only the PnH DLC BGAs, as you suspected..

**Comment by Mmtrx** [comment]
Could you also test a slightly better version, I don't have the PnH dlc? (SILAGE could be a different price from GRASS_FERMENTED):
```
	-- check for Maize+ filltype
	local fillType = m.fillType
	if m.sellPoint.fillTypePrices[fillType] == nil then
		fillType = FillType.SILAGE
	end
	return m.sellPoint:getEffectiveFillTypePrice(fillType)
```
and in gui.lua you can just say
```
		--get current price
		local price = self:getFilltypePrice(m)
```
(This time, small t is ok 😉)

**Comment by HiPhi77** [comment]
changed it to your last version. Works fine. (this time also tested on a windows version of the game, just to be sure)
and yes, GRASS_FERMENTED and SILAGE can have a different price. Just checked, in my test save i see 153 for SILAGE but 134 for GRASS_FERMENTED. Better contracts displays 153 for the bale mission.
But I think this is fine, as GRASS_FERMENTED is converted to SILAGE when filling to a PnH BGA - according to docu it converts 1000l GRASS_FERMENTED to 980l SILAGE. So there might be a small error in the expected sell gain in better contracts, in my opinion this can be ignored :)

**Comment by Mmtrx** [comment]
Great, was just released as part of [v1.2.7.3](https://github.com/Mmtrx/FS22_BetterContracts/releases/tag/v1.2.7.3).
